test/qa/test_meshnet.py

```diff
@@ -170,3 +196,120 @@ def test_set_meshnet_off_repeated(meshnet_allias):
+
+    daemon.stop_peer(ssh_client)
+    ssh_client.exec_command("echo '' > /var/log/nordvpn/daemon.log")
+    daemon.start_peer(ssh_client)
+
+    time.sleep(30)
+
+    dlog = ssh_client.exec_command("cat /var/log/nordvpn/daemon.log").split("\n")
+
+    derp_lines_from_logs =  meshnet.get_lines_with_keywords(dlog, ["body", "relay", "connected"])
+
+    server_list = []
+    for line in derp_lines_from_logs:
+        json_match = re.search(r'\{.*\}', line)
+
+        if json_match:
+            json_data = json.loads(json_match.group())
+            hostname = json_data.get("body", {}).get("hostname", "")
+            server_list.append(hostname)
+
+    # Same server should not be contacted twice in a row
+    assert not has_duplicates(server_list)
+
+    ssh_client.exec_command("sudo iptables -D OUTPUT -p tcp -m tcp --sport 8765 -j DROP")
+    ssh_client.exec_command("sudo iptables -D OUTPUT -p tcp -m tcp --dport 8765 -j DROP")
+
+
+@pytest.mark.skip("LVPN-3428, need a discussion here")
+@pytest.mark.flaky(reruns=2, reruns_delay=90)
+@timeout_decorator.timeout(80)
+def test_direct_connection_rtt_and_loss():
+    def get_loss(ping_output: str) -> float:
+        """ pass `ping_output`, and get loss returned. """
+        return float(ping_output.split("\n")[-3].split(", ")[2].split("%")[0])
+
+    def get_average_rtt(ping_output: str) -> float:
+        """ pass `ping_output`, and get average rtt returned. """
+        return float(ping_output.split("\n")[-2].split("/")[4])
+
+    def base_test(log: str, peer_hostname: str):
+        log = log.split("\n")
+        log_relay_events = meshnet.get_lines_with_keywords(log, [peer_hostname, "body", "relay", "connected"])
+        log_direct_events = meshnet.get_lines_with_keywords(log, [peer_hostname, "body", "direct", "connected"])
+
+        # Sometimes no relay lines show up, but direct ones do instead. If that happens, direct formed, so we can continue
+        if len(log_relay_events) != 0:
+            log_relay_event_time = datetime.strptime(log_relay_events[0].split(" ")[1], "%H:%M:%S")
+            log_direct_event_time = datetime.strptime(log_direct_events[0].split(" ")[1], "%H:%M:%S")
+
+            assert (log_direct_event_time - log_relay_event_time).total_seconds() < meshnet.TELIO_EXPECTED_RELAY_TO_DIRECT_TIME
+        elif len(log_relay_events) == 0 and len(log_direct_events) > 0:
+            pass
+
+        # RTT & loss
+        ping_output = sh.ping("-c", "20", peer_hostname)
+        assert get_average_rtt(ping_output) <= meshnet.TELIO_EXPECTED_RTT
+        assert get_loss(ping_output) <= meshnet.TELIO_EXPECTED_PACKET_LOSS
+
+    peer_list = meshnet.PeerList.from_str(sh.nordvpn.mesh.peer.list())
+
+    with open(logging.FILE) as tester_log:
+        log_content = tester_log.read()
+        qapeer_hostname = peer_list.get_external_peer().hostname
+        base_test(log_content, qapeer_hostname)
+
+
+@timeout_decorator.timeout(40)
```

Comment:
why is needed to fail after 40 seconds?

Reply:
Good point. For me, it has became pretty normal to add this decorator to almost each test :smile: